optuna/_hypervolume/box_decomposition.py

```diff
@@ -0,0 +1,144 @@
+"""
+The functions in this file are mostly based on BoTorch v0.13.0,
+but they are refactored significantly from the original version.
+
+For ``_get_upper_bound_set``, look at:
+    * https://github.com/pytorch/botorch/blob/v0.13.0/botorch/utils/multi_objective/box_decompositions/utils.py#L101-L160
+
+For ``_get_box_bounds``, look at:
+    * https://github.com/pytorch/botorch/blob/v0.13.0/botorch/utils/multi_objective/box_decompositions/utils.py#L163-L193
+
+For ``_get_non_dominated_box_bounds``, look at:
+    * https://github.com/pytorch/botorch/blob/v0.13.0/botorch/utils/multi_objective/box_decompositions/non_dominated.py#L395-L430
+
+The preprocessing for four or fewer objectives, we use the algorithm proposed by:
+    Title: A Box Decomposition Algorithm to Compute the Hypervolume Indicator
+    Authors: Renaud Lacour, Kathrin Klamroth, and Carlos M. Fonseca
+    URL: https://arxiv.org/abs/1510.01963
+We refer this paper as Lacour17 in this file.
+
+"""  # NOQA: E501
+
+from __future__ import annotations
+
+import warnings
+
+import numpy as np
+
+from optuna.study._multi_objective import _is_pareto_front
+
+
+def _get_upper_bound_set(
+    sorted_pareto_sols: np.ndarray, ref_point: np.ndarray
+) -> tuple[np.ndarray, np.ndarray]:
+    """
+    This function follows Algorithm 2 of Lacour17.
+
+    Args:
+        sorted_pareto_sols: Pareto solutions sorted with respect to the first objective.
+        ref_point: The reference point.
+
+    Returns:
+        upper_bound_set: The upper bound set, which is ``U(N)`` in the paper. The shape is
+        ``(n_bounds, n_objectives)``.
+        def_points: The defining points of each vector in ``U(N)``. The shape is
+        ``(n_bounds, n_objectives, n_objectives)``.
+
+    NOTE:
+        ``pareto_sols`` corresponds to ``N`` and ``upper_bound_set`` corresponds to ``U(N)`` in the
+        paper.
+        ``def_points`` (the shape is ``(n_bounds, n_objectives, n_objectives)``) is not well
+        explained in the paper, but basically, ``def_points[i, j] = z[j]`` of
+        ``upper_bound_set[i]``.
+    """
+    (_, n_objectives) = sorted_pareto_sols.shape
+    objective_indices = np.arange(n_objectives)
+    skip_ineq_judge = np.eye(n_objectives, dtype=bool)
+    # NOTE(nabenabe): False at 0 comes from Line 2 of Alg. 2. (loss_vals is sorted w.r.t. 0-th obj)
+    skip_ineq_judge[:, 0] = True
+
+    def update(sol: np.ndarray, ubs: np.ndarray, dps: np.ndarray) -> tuple[np.ndarray, np.ndarray]:
+        # The update rule is written in Section 2.2 of Lacour17.
+        is_dominated = np.all(sol < ubs, axis=-1)
```

Comment:
In the original paper, they adopt the weekly dominates `<=`. Should we use the same one? If we want to make it tolerant to numerical errors, how about using something like `sol <= ubs - eps`?

Reply:
It appears they use strong dominance, don't they?

![image](https://github.com/user-attachments/assets/17860bab-8c1a-4b88-a64f-23a2473473ef)

As we use `sol - ubs` in the hypervolume calculation and it may potentially increase the computational overhead, it would be better to avoid using the condition that makes `sol - ubs` negative and increases the size of `A` in the paper.

Reply:
Oh, sorry for confusion. I misunderstood.